core/src/main/java/com/linecorp/armeria/common/stream/TimeoutStreamMessage.java

```diff
@@ -0,0 +1,273 @@
+    @Override
+    public long demand() {
+        return delegate.demand();
+    }
+
+    /**
+     * Returns a {@link CompletableFuture} that completes when this stream is complete,
+     * either successfully or exceptionally.
+     *
+     * @return a future that completes when the stream is complete
+     */
+    @Override
+    public CompletableFuture<Void> whenComplete() {
+        return delegate.whenComplete();
+    }
+
+    /**
+     * Creates a {@link TimeoutSubscriber} with timeout logic applied using the specified
+     * subscriber, executor, duration, and mode.
+     * Then subscribes to the original stream with the TimeoutSubscriber, executor, and options.
+     *
+     * @param subscriber the subscriber to this stream
+     * @param executor the executor for running timeout tasks and stream operations
+     * @param options subscription options
+     * @see StreamMessage#subscribe(Subscriber, EventExecutor, SubscriptionOption...)
+     */
+    @Override
+    public void subscribe(Subscriber<? super T> subscriber, EventExecutor executor,
+                          SubscriptionOption... options) {
+        timeoutSubscriber = new TimeoutSubscriber<>(subscriber, executor, timeoutDuration, timeoutMode);
+        delegate.subscribe(timeoutSubscriber, executor, options);
+    }
+
+    /**
+     * Cancels the timeout schedule if it is set.
+     */
+    private void cancelSchedule() {
+        if (timeoutSubscriber != null) {
+            timeoutSubscriber.cancelSchedule();
+        }
+    }
+
+    /**
+     * Aborts the stream with an {@link AbortedStreamException} and prevents further subscriptions.
+     * Calling this method on a closed or aborted stream has no effect.
+     * Cancels the timeout schedule if it is set.
+     */
+    @Override
+    public void abort() {
+        cancelSchedule();
+        delegate.abort();
+    }
+
+    /**
+     * Aborts the stream with the specified {@link Throwable} and prevents further subscriptions.
+     * Calling this method on a closed or aborted stream has no effect.
+     * Cancels the timeout schedule if it is set.
+     *
+     * @param cause the cause of the abort
+     */
+    @Override
+    public void abort(Throwable cause) {
+        cancelSchedule();
+        delegate.abort(cause);
+    }
+
+    static final class TimeoutSubscriber<T> implements Runnable, Subscriber<T>, Subscription {
+
+        private static final String TIMEOUT_MESSAGE = "Stream timed out after %d ms (timeout mode: %s)";
+        private final Subscriber<? super T> delegate;
+        private final EventExecutor executor;
+        private final StreamTimeoutMode timeoutMode;
+        private final Duration timeoutDuration;
+        private final long timeoutNanos;
+        @Nullable
+        private ScheduledFuture<?> timeoutFuture;
+        @Nullable
+        private Subscription subscription;
+        private long lastEventTimeNanos;
+        private boolean completed;
+        private boolean canceled;
```

Comment:
`cancel()` may be called in a different thread so `canceled=true` could be invisible in `onNext()`.
```suggestion
        private volatile boolean canceled;
```